src/NServiceBus.Core/Persistence/Learning/SagaPersister/LearningSynchronizedStorageSession.cs

```diff
@@ -0,0 +1,90 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Collections.Generic;
+    using System.Threading.Tasks;
+    using Janitor;
+    using Persistence;
+    using Sagas;
+
+    [SkipWeaving]
+    class LearningSynchronizedStorageSession : CompletableSynchronizedStorageSession
+    {
+        public LearningSynchronizedStorageSession(SagaManifestCollection sagaManifests)
+        {
+            this.sagaManifests = sagaManifests;
+        }
+
+        public void Dispose()
+        {
+            foreach (var sagaFile in sagaFiles.Values)
+            {
+                sagaFile.Dispose();
+            }
+
+            sagaFiles.Clear();
+        }
+
+        public async Task CompleteAsync()
+        {
+            foreach (var action in deferredActions)
+            {
+                await action.Execute().ConfigureAwait(false);
+            }
+            deferredActions.Clear();
+        }
+
+        public Task<TSagaData> Read<TSagaData>(Guid sagaId) where TSagaData : IContainSagaData
+        {
+            return Task.Run(() =>
```

Comment:
The initial thought was that the file creation and serialization can take a while and is completely synchronous due to how file access works. Happy to remove